src/XMakeBuildEngine/Definition/ToolsetConfigurationReader.cs

```diff
@@ -245,46 +245,39 @@ protected override IEnumerable<ToolsetPropertyDefinition> GetSubToolsetPropertyD
                 return kindToPathsCache;
             }
 
-            var allPaths = new MSBuildExtensionsPathReferenceKind[] {
-                                MSBuildExtensionsPathReferenceKind.Default,
-                                MSBuildExtensionsPathReferenceKind.Path32,
-                                MSBuildExtensionsPathReferenceKind.Path64,
-                                MSBuildExtensionsPathReferenceKind.None
-                            };
-
-            foreach (MSBuildExtensionsPathReferenceKind kind in allPaths)
-            {
-                kindToPathsCache[kind] = ComputeDistinctListOfFallbackPathsFor(kind, propertyCollection);
-            }
+            kindToPathsCache = ComputeDistinctListOfFallbackPathsFor(propertyCollection);
 
             return kindToPathsCache;
         }
 
         /// <summary>
         /// Returns a list of the search paths for a given MSBuildExtensionsPathReferenceKind
         /// </summary>
-        private List<string> ComputeDistinctListOfFallbackPathsFor(MSBuildExtensionsPathReferenceKind refKind, ToolsetElement.PropertyElementCollection propertyCollection)
+        private Dictionary<string, List<string>> ComputeDistinctListOfFallbackPathsFor(ToolsetElement.PropertyElementCollection propertyCollection)
         {
-            var extnPaths = new List<string>();
-            var pathsFromConfig = propertyCollection.GetElement(refKind.StringRepresentation)?.Value;
+            var paths = new Dictionary<string, List<string>>();
 
-            //FIXME: handle ; in path on Unix
-            if (String.IsNullOrEmpty(pathsFromConfig))
+            foreach (ToolsetElement.PropertyElement property in propertyCollection)
             {
-                return extnPaths;
-            }
+                //FIXME: handle ; in path on Unix
+                if (string.IsNullOrEmpty(property.Name) || string.IsNullOrEmpty(property.Value))
```

Comment:
Maybe we can also verify the property name is valid with `MSBuildExtensionsPathReferenceKind.FindIn` ?